compiler/rustc_errors/src/emitter.rs

```diff
@@ -2557,62 +2557,60 @@ fn num_decimal_digits(num: usize) -> usize {
     MAX_DIGITS
 }
 
-// We replace some characters so the CLI output is always consistent and underlines aligned.
-// Keep the following list in sync with `rustc_span::char_width`.
-const OUTPUT_REPLACEMENTS: &[(char, &str)] = &[
-    ('\t', "    "),    // We do our own tab replacement
-    ('\u{200D}', ""), // Replace ZWJ with nothing for consistent terminal output of grapheme clusters.
-    ('\u{202A}', "�"), // The following unicode text flow control characters are inconsistently
-    ('\u{202B}', "�"), // supported across CLIs and can cause confusion due to the bytes on disk
-    ('\u{202D}', "�"), // not corresponding to the visible source code, so we replace them always.
-    ('\u{202E}', "�"),
-    ('\u{2066}', "�"),
-    ('\u{2067}', "�"),
-    ('\u{2068}', "�"),
-    ('\u{202C}', "�"),
-    ('\u{2069}', "�"),
-    // In terminals without Unicode support the following will be garbled, but in *all* terminals
-    // the underlying codepoint will be as well. We could gate this replacement behind a "unicode
-    // support" gate.
-    ('\u{0000}', "␀"),
-    ('\u{0001}', "␁"),
-    ('\u{0002}', "␂"),
-    ('\u{0003}', "␃"),
-    ('\u{0004}', "␄"),
-    ('\u{0005}', "␅"),
-    ('\u{0006}', "␆"),
-    ('\u{0007}', "␇"),
-    ('\u{0008}', "␈"),
-    ('\u{000B}', "␋"),
-    ('\u{000C}', "␌"),
-    ('\u{000D}', "␍"),
-    ('\u{000E}', "␎"),
-    ('\u{000F}', "␏"),
-    ('\u{0010}', "␐"),
-    ('\u{0011}', "␑"),
-    ('\u{0012}', "␒"),
-    ('\u{0013}', "␓"),
-    ('\u{0014}', "␔"),
-    ('\u{0015}', "␕"),
-    ('\u{0016}', "␖"),
-    ('\u{0017}', "␗"),
-    ('\u{0018}', "␘"),
-    ('\u{0019}', "␙"),
-    ('\u{001A}', "␚"),
-    ('\u{001B}', "␛"),
-    ('\u{001C}', "␜"),
-    ('\u{001D}', "␝"),
-    ('\u{001E}', "␞"),
-    ('\u{001F}', "␟"),
-    ('\u{007F}', "␡"),
-];
-
 fn normalize_whitespace(str: &str) -> String {
-    let mut s = str.to_string();
-    for (c, replacement) in OUTPUT_REPLACEMENTS {
-        s = s.replace(*c, replacement);
-    }
-    s
+    // We replace some characters so the CLI output is always consistent and underlines aligned.
```

Comment:
Can you put this in a `LazyCell` or whatever?

Reply:
sorting it and using binary search would be another option

Reply:
Does anyone have thoughts about how to avoid the `String` allocation? Doubt it actually affects performance that much... but still.

Reply:
You can return a borrowed `Cow` if no substitutions happened and an owned one if one does happen.

Reply:
And a big `match` might be even simpler.

Reply:
@the8472 what do you think of the following?

```rust
    str.chars()
        .flat_map(|c| match output_replacements.binary_search_by_key(&c, |(k, _)| *k) {
            Ok(i) => Either::Left(output_replacements[i].1.chars()),
            _ => Either::Right([c].into_iter()),
        })
        .collect()
```

There would be a single heap allocation by `collect` and nothing else, right?

Reply:
Well, it doesn't know the length so it'd still do the amortized growth thing, so `O(log(l))` allocations.
And it wouldn't avoid an allocation if nothing changed.

But it's better than what's currently in the PR.